File: cos/olympus/src/main/java/cos/olympus/game/Damages.java
package cos.olympus.game;

import cos.logging.Logger;
import cos.olympus.game.events.Damage;
import cos.olympus.game.events.Spell;

import java.util.ArrayList;
import java.util.function.Consumer;

public class Damages implements TickAware {
    private static int DAMAGES_IDS = 0;
    private final static Logger logger = Logger.get(Damages.class);
    private ArrayList<Damage> data = new ArrayList<>();
    private int tick;

    @Override
    public void onTick(int tick) {
        this.tick = tick;
    }

    public void on(Creature victim, Spell spell, int amount, boolean crit) {
        var dmg = new Damage(++DAMAGES_IDS, tick, victim, spell, amount, crit);
        logger.info(dmg);
        data.add(dmg);
    }

    public void forEach(Consumer<Damage> consumer) {
        if (data.isEmpty()) return;

        data.forEach(consumer);
    }

    public void clear() {
        if (data.isEmpty()) return;

        data = new ArrayList<>();
    }
}

File: cos/src/main/java/cos/olympus/game/Damages.kt
package cos.olympus.game

import cos.olympus.game.events.Damage
import cos.olympus.game.events.Spell
import java.util.function.Consumer

class Damages : TickAware {
    private var data = ArrayList<Damage>()
    private var tick = 0

    override fun onTick(tick: TickId) {
        this.tick = tick
    }

    fun on(victim: Actor, spell: Spell, amount: Int, crit: Boolean) {
        val dmg = Damage(++DAMAGES_IDS, tick, victim, spell, amount, crit)
        ///        logger.info(dmg);
        data.add(dmg)
    }

    fun forEach(consumer: Consumer<Damage>?) {
        if (data.isEmpty()) return

        data.forEach(consumer)
    }

    fun clear() {
        if (data.isEmpty()) return

        data = ArrayList()
    }

    companion object {
        private var DAMAGES_IDS = 0
    }
}
